LSPS0/common-schemas.md

```diff
@@ -0,0 +1,245 @@
+>
+> We expect that proportional amounts would be smaller than
+> 100% or 1.0, which would be encoded as the JSON integral
+> number 1000000, which is small enough to easily fit into
+> IEEE 754 numbers or 32-bit signed integers with no loss
+> of significant bits.
+
+### Short Channel Identifiers (SCID)
+
+SCIDs MUST be encoded as a JSON string containing the
+"human-readable" format of `BBBxTTTxOOO`.
+
+`BBB` is the top 24 bits in decimal text.
+`TTT` is the middle 24 bits in decimal text.
+`OOO` is the lowest 16 bits in decimal text.
+`x` are literal lowercase `x` characters.
+
+For example, an SCID which would be hex-dumped as the binary
+blob `083a8400034d0001` when encoded in a typical BOLT binary
+encoding, would be written as the JSON string `"539268x845x1"`.
+
+> **Rationale** This format is a recognizable and distinctive
+> format for SCIDs, and helps separate the SCID type from other
+> types.
+
+### SECP256K1 Points / Public Keys / Lightning Network Node IDs
+
+Lightning Network node IDs are SECP256K1 ECC public keys, which
+are points on the SECP256K1 elliptic curve.
+
+SECP256K1 elliptic curve points MUST be encoded as a JSON string
+of the hexadecimal dump of the compressed DER encoding of the
+point.
+
+The compressed DER encoding is a 33-byte representation, and the
+JSON string hexadecimal dump would therefore be encoded as 66
+hexadecimal characters.
+The first byte is either the byte `0x02` for an even Y coordinate,
+or `0x03` for an odd Y coordinate, while the rest of the bytes is
+the big-endian 32-byte integer representation of the X coordinate.
+
+In contexts where the case is significant (for example, if it
+will be committed to by some hash or signature) then the
+representation MUST be in lowercase.
+Otherwise, readers MUST allow both lowercase and uppercase
+hexadecimal digits.
+
+Readers SHOULD validate that the point is indeed on the SECP256K1
+curve.
+
+For example, the SECP256K1 generator point `G` would be written
+as the JSON string
+`"0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798"`.
+
+> **Rationale** Bitcoin and Lightning have standardized on this
+> format, and traditionally used lowercase for hexadecimal digits.
+
+### Lightning Network Connection Strings
+
+Lightning Network connection strings describe a Lightning Network
+Node ID and one way to connect to that node.
+
+Connection strings MUST be encoded as a JSON string, with three
+parts:
+
+1.  The Lightning Network Node ID, encoded as the hexadecimal
+    dump of the compressed DER encoding, followed by the `@`
+    sign.
+2.  The address, which follows the `@` sign and lasts to the
+    end of the string, or the last `:` in the string, or
+    delimited by characters `[` and `]`.
+    Address formats allowed are those defined in the BOLT
+    specifications:
+    * IPv4 addresses.
+    * IPv6 addresses.
+    * A DNS-resolvable name.
+    * A Torv3 hidden service.
+3.  An optional port number, which is started by the last `:`
```

Comment:
Making port number mandatory also means `[` `]` delimiters for the address are unnecessary, as e.g. `<node>@::1:9735` unambiguously means that the address is `::1` since the last `:` is the indicator for the port always.  This is an advantage.  On the other hand everyone is going to be on `:9735` in practice.  Hmm

Reply:
Nobody will type this manually. It will be inserted either by a wallet itself (that has to specific the port anyway because it can't assume it's always 9735) or copy-paste it from an explorer (which adds the port by default). There is no use case where the client only adds a pubkey@address.

Reply:
Okay, make port mandatory and then remove `[` `]` option seems good to me.

Reply:
Done.